### PR TITLE
feat: Add retry logic for PubSub and Implement Deserialization generalized

### PR DESCRIPTION
**Pull request by XavierGeerinck**
# Description

- Introduced Deserializer and corresponding tests
- Added DROP/RETRY/SUCCESS Logic for Pub/Sub and corresponding tests

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #338 and #452 

## To Discuss

- [ ] Topics in raw payload should be handled as raw payload and not decoded. Currently when we send a string ("Hello World") dapr will pass it through as text/plain but we receive it as application/octet-stream. We decode this now and return the string, but this will break
- [ ] Do dropped messages land in deadletter?

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Extended the documentation

### CODE REVIEW

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 1]
```diff
@@ -228,6 +228,49 @@ async function start() {
...
+  const pubSubName = "my-pubsub-name";
+  const topic = "topic-a";
+
+  // Process a message successfully
+  await server.pubsub.subscribe(pubSubName, topic, async (data: any, headers: object) => {
+    return DaprPubSubStatusEnum.SUCCESS;
+  });
+
+  // Retry a message
+  // Note: this example will keep on retrying to deliver the message
```
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:259]: Should we add a note that this is component specific?

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 2]
```diff
@@ -228,6 +228,49 @@ async function start() {
...
+  await server.pubsub.subscribe(pubSubName, topic, async (data: any, headers: object) => {
+    return DaprPubSubStatusEnum.RETRY;
+  });
+
+  // Drop a message
+  await server.pubsub.subscribe(pubSubName, topic, async (data: any, headers: object) => {
+    return DaprPubSubStatusEnum.DROP;
```
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:266]: What is the default behavior if users don't return anything? Should we document it here as well? Or we can provide a `Promise` style syntax where `resolve` always take in an enum input.

File: test/e2e/common/server.test.ts [hunk 3]
```diff
@@ -47,6 +48,30 @@ describe("common/server", () => {
...
+      //   mockSubscribeStatusHandlerRetryCounterIsEnabled = true;
+
+      //   if (mockSubscribeStatusHandlerRetryCounter < 2) {
+      //     mockSubscribeStatusHandlerRetryCounter++;
+      //     return DaprPubSubStatusEnum.RETRY;
+      //   }
+      default:
+        // mockSubscribeStatusHandlerRetryCounterIsEnabled = false;
+        // mockSubscribeStatusHandlerRetryCounter = 0;
+        return DaprPubSubStatusEnum.SUCCESS;
```
Comment (shubham1172) [inline test/e2e/common/server.test.ts:71]: TODO?

File: package.json [hunk 4]
```diff
@@ -43,6 +43,7 @@
     "@types/google-protobuf": "^3.15.5",
     "@types/node-fetch": "^2.6.2",
     "body-parser": "^1.19.0",
+    "cloudevents": "^6.0.3",
```
Comment (shubham1172) [inline package.json:46]: Do we need to add this dependency? I propose that we stick to "required" fields on CloudEvents to check for it, and return a simple object, letting users cast it to CloudEvent objects if required.

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 5]
```diff
@@ -436,19 +433,67 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+      let status = DaprPubSubStatusEnum.SUCCESS;
+
+      try {
+        status = await cb(data, headers);
+      } catch (e) {
+        // We catch and log an error, but we don't do anything with it as the statuses should define that
+        this.logger.error(`[route-${routeObj.path}] Message processing failed, ${e}`);
+
+        if (status === DaprPubSubStatusEnum.SUCCESS) {
+          status = DaprPubSubStatusEnum.RETRY;
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:476]: TODO: FIX and Drop the message

File: src/implementation/Server/HTTPServer/HTTPServerImpl.ts [hunk 6]
```diff
@@ -146,6 +139,46 @@ export default class HTTPServerImpl {
...
+      let status = DaprPubSubStatusEnum.SUCCESS;
+
+      try {
+        status = await cb(data, headers);
+      } catch (e) {
+        // We catch and log an error, but we don't do anything with it as the statuses should define that
+        this.logger.error(`[route-${routeObj.path}] Message processing failed, ${e}`);
+
+        if (status === DaprPubSubStatusEnum.SUCCESS) {
+          status = DaprPubSubStatusEnum.RETRY;
```
Comment (XavierGeerinck) [inline src/implementation/Server/HTTPServer/HTTPServerImpl.ts:161]: @todo: remove

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 1 above)
Comment (XavierGeerinck) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:259]: How do you mean component specific?

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 2 above)
Comment (XavierGeerinck) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:266]: SUCCESS is returned by default but a message is logged when processing fails

File: package.json (hunk 4 above)
Comment (XavierGeerinck) [reply package.json:46]: Removed it :)

**Comment by XavierGeerinck** [comment]
image

Stuck on the above, the tests do seem to run in async? @shubham1172 the `runIt` seem to trigger HTTP and gRPC at the same time

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 1 above)
Comment (shubham1172) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:259]: So components can have their own retry configurations, e.g., https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-redis-pubsub/, `redisMaxRetries` is 0 by default.

File: package.json (hunk 4 above)
Comment (shubham1172) [reply package.json]: Thank you!

**Comment by shubham1172** [comment]
> @shubham1172 the runIt seem to trigger HTTP and gRPC at the same time

It just triggers `it()` for HTTP and gRPC, it can overlap AFAIK. If you use `protocol` in the topic queue name, then you can separate HTTP and gRPC. Will it still be a problem?

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 1 above)
Comment (XavierGeerinck) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:259]: Ok added a note

**Comment by XavierGeerinck** [comment]
@shubham1172 :D

**Comment by shubham1172** [comment]
Conflicts due to merge 😛 @XavierGeerinck

**Review by shubham1172 — APPROVED** [review]

**Pull request merged**